### ISSUE TITLE
Facet documentation is incorrect

### ISSUE THREAD

**Issue by pjjanak** [algolia/algoliasearch-client-java#526]
- Algolia Client Version: 2.21.0
- Language Version: 1.8.0_162

### Description
The documentation for faceting is incorrect, however I think the actual problem lies in the implementation. [This](https://www.algolia.com/doc/api-reference/api-parameters/facets/) states that you can supply facets as an Array, which is what the Algolia backend seems to expect (eg, `facets=["foo","bar"]`). 

However, `QueryBase` defines facets as being a String, which when serialized will produce `facets="foo"`.

This also breaks seamless integration with the instant-search library, as it sends facets as a JSON Array to whatever backend is specified. If we use this library as our backend it will fail when attempting to facet. One must intercept the output of react-instantsearch before the query is submitted to the backend and modify the facets parameter so it is a String.

**Comment by Ant-hem** [comment]
Hello!

Thanks again for your contribution. 

The doc is explaining the "new" way of passing `facets` (as string[]). However, the engine is backward compatible with the "old" way of passing `facets` (i.e `"foo,bar"`. If you are in a hurry you can pass your facets separated by commas in a string it will work. 

I will add the support for `string[]` but I need to do it differently that your proposal because it's a BC break. (changing existing signature)

**Comment by pjjanak** [comment]
Makes sense - I'm also realizing that it seems certain requests expect Strings still - for instance the requests to get the number of results that match a facet. 

Maybe my bug is actually a different one then. When I try to do a faceted query I get back 0 results, despite knowing that there are results. The aforementioned query to get the number of results matching a facet informs me that I should be seeing 723 , but I get 0 nbHits in the "real" request. Here are my requests and the respective responses:

```
{
   "requests":[
      {
         "indexName":"Dev-MediaNews",
         "params":{
            "query":"",
            "maxValuesPerFacet":10,
            "page":0,
            "highlightPreTag":"<ais-highlight-0000000000>",
            "highlightPostTag":"</ais-highlight-0000000000>",
            "filters":"publishDate<=1545062453000",
            "facetingAfterDistinct":true,
            "facets":"type",
            "facetFilters":[
               [
                  "type:News Release"
               ]
            ]
         }
      },
      {
         "indexName":"Dev-MediaNews",
         "params":{
            "query":"",
            "maxValuesPerFacet":10,
            "page":0,
            "highlightPreTag":"<ais-highlight-0000000000>",
            "highlightPostTag":"</ais-highlight-0000000000>",
            "filters":"publishDate<=1545062453000",
            "facetingAfterDistinct":true,
            "hitsPerPage":1,
            "attributesToRetrieve":[],
            "attributesToHighlight":[],
            "attributesToSnippet":[],
            "analytics":false,
            "clickAnalytics":false,
            "facets":"type"
         }
      }
   ]
}
```

```
{
   "results":[
      {
         "page":0,
         "nbHits":0,
         "nbPages":0,
         "hitsPerPage":20,
         "processingTimeMS":1,
         "facets":{},
         "facets_stats":null,
         "exhaustiveFacetsCount":true,
         "query":"",
         "params":"filters=publishDate%3C%3D1545062453000&facets=type&maxValuesPerFacet=10&facetFilters=%5B%5Btype%3ANews+Release%5D%5D&facetingAfterDistinct=true&highlightPreTag=%3Cais-highlight-0000000000%3E&highlightPostTag=%3C%2Fais-highlight-0000000000%3E&page=0&query=",
         "hits":[],
         "index":"Dev-MediaNews",
         "processed":null,
         "queryID":null,
         "userData":null,
         "appliedRules":null
      },
      {
         "page":0,
         "nbHits":891,
         "nbPages":891,
         "hitsPerPage":1,
         "processingTimeMS":1,
         "facets":{
            "type":{
               "News Release":723,
               "Media Advisory":134,
               "Transcript":34
            }
         },
         "facets_stats":null,
         "exhaustiveFacetsCount":true,
         "query":"",
 
 
 "params":"analytics=false&clickAnalytics=false&attributesToRetrieve=&filters=publishDate%3C%3D1545062453000&facets=type&maxValuesPerFacet=10&facetingAfterDistinct=true&attributesToHighlight=&attributesToSnippet=&highlightPreTag=%3Cais-highlight-0000000000%3E&highlightPostTag=%3C%2Fais-highlight-0000000000%3E&page=0&hitsPerPage=1&query=",
         "hits":[
            {
               "objectID":"12469-2"
            }
         ],
         "index":"Dev-MediaNews",
         "processed":null,
         "queryID":null,
         "userData":null,
         "appliedRules":null
      }
   ]
}
```

**Comment by pjjanak** [comment]
In the meantime, just supplying a filter string with the facet values I want to search on seems to be working fine. So I do have a working solution at the moment. 

eg,

```
filters: "publishDate<=1545068512272 AND (type:"News Release")"
```

in my request.

**Comment by Ant-hem** [comment]
Hey!

What do you want achieve exactly?

 If you want to **filter** your query this is the right way. Faceting is more used when you want to build a UI. The two concepts overlap and are often confused (they are closely linked). Could you please read [the faceting overview ](https://www.algolia.com/doc/guides/searching/faceting/#faceting-overview) and tell me if it answers to you? It will explain better than me 😄

**Comment by pjjanak** [comment]
Yeah - I think there's some confusion with my terminology haha - but I believe what I was trying to do initially is still not working correctly with the above request. Let me try and break it down better.

1. I am attempting to **facet** the data set. This is working correctly.
2. I am attempting **filter on facets** of the data set. This is what is not working if I attempt to supply facet filters via the `facetFilter` parameter. If, however, I supply the entirety of my filters (of which I have a numeric filter on `publishDate` and 1 or more facet filters on `type`) then I get the data I would expect back.

Additionally, for more information on what I was trying to do specifically: I was trying to use the `RefinementList` of React InstantSearch ([Link](https://community.algolia.com/react-instantsearch/widgets/RefinementList.html)). I am providing a custom backend to InstantSearch which points to an endpoint which uses this Java client. We want to do this for security purposes as the contents of this data set are private and a user must be authenticated before they have access to it.

The `RefinementList` will cause the InstantSearch to both facet and filter on facets. It's method of filtering on facets is to specify a `facetFilters` parameter as a JSON array of Strings (hence the original idea behind this issue).

For the moment I have moved away from the `RefinementList` entirely and am providing a custom UI which will update the `filters`parameter of the search request as described above. This works for me since the "refinement values" for this data set are well defined. But for someone who would like to use the `RefinementList` out of the box I believe this client will currently fail.

**Comment by Ant-hem** [comment]
Thanks for the explanation. I now understand what you want to achieve. InstantSearch is a bit different than Java/C#/Go etc.. library because InstantSearch is closely linked to the UI. So it provides lots of feature "backend" libraries don't have. 

However, I'll still fix this issue because we are not yet supporting the new `facets` structure.

**Comment by Ant-hem** [comment]
Fixed with v3.0.0 (febbaea)
